File: src/ops/FlashDrum.cpp
#include "chemsim/ops/FlashDrum.hpp"
#include <stdexcept>

namespace chemsim {

FlashDrum::FlashDrum(Spec spec, const FlashCalculator& fc)
    : spec_(spec), fc_(fc) {}

void FlashDrum::populateOutlets(const FlashResult& r) {
    vapor.T = liquid.T = r.T;
    vapor.P = liquid.P = r.P;
    vapor.vaporFraction  = 1.0;
    liquid.vaporFraction = 0.0;
    vapor.phase  = Phase::VAPOR;
    liquid.phase = Phase::LIQUID;

    vapor.z  = vapor.y  = vapor.x  = r.y;
    liquid.z = liquid.x = liquid.y = r.x;

    vapor.totalFlow  = feed.totalFlow * r.beta;
    liquid.totalFlow = feed.totalFlow * (1.0 - r.beta);

    // Single-phase enthalpies for each outlet stream
    vapor.H  = fc_.phaseEnthalpy(r.T, r.P, r.y, false);
    liquid.H = fc_.phaseEnthalpy(r.T, r.P, r.x, true);
    vapor.S  = fc_.phaseEntropy (r.T, r.P, r.y, false);
    liquid.S = fc_.phaseEntropy (r.T, r.P, r.x, true);

    // If fully single-phase, one outlet gets zero flow
    if (r.beta >= 1.0 - 1e-10) {
        vapor.totalFlow  = feed.totalFlow;
        liquid.totalFlow = 0.0;
        liquid.z = liquid.x = liquid.y = r.y;  // copy composition for zero-flow stream
    } else if (r.beta <= 1e-10) {
        liquid.totalFlow = feed.totalFlow;
        vapor.totalFlow  = 0.0;
        vapor.z = vapor.y = vapor.x = r.x;
    }
}

void FlashDrum::solve() {
    if (feed.totalFlow <= 0.0)
        throw std::invalid_argument("FlashDrum: feed.totalFlow must be > 0");
    if (feed.z.empty())
        throw std::invalid_argument("FlashDrum: feed.z is empty");

    // Compute inlet enthalpy (needed for duty in TP mode)
    // Use flash at feed conditions to get H_in
    auto r_in = fc_.flashTP(feed.T, feed.P, feed.z);
    double H_in = fc_.totalEnthalpy(r_in);

    FlashResult r_out;
    switch (spec_) {
    case Spec::TP:
        r_out = fc_.flashTP(T_spec, P_spec, feed.z);
        break;
    case Spec::PH:
        r_out = fc_.flashPH(P_spec, H_spec, feed.z, feed.T);
        break;
    case Spec::PS:
        r_out = fc_.flashPS(P_spec, S_spec, feed.z, feed.T);
        break;
    }

    populateOutlets(r_out);

    double H_out = fc_.totalEnthalpy(r_out);
    duty = feed.totalFlow * (H_out - H_in);   // [W]
}

} // namespace chemsim

File: chemsim/ops/flash_drum.py
"""Two-outlet flash separator (TP, PH, or PS specification)."""
from __future__ import annotations

from enum import Enum
from typing import Optional

from chemsim.core import Phase, Stream
from chemsim.ops.base import IUnitOp
from chemsim.thermo.flash import FlashCalculator, FlashResult


class FlashSpec(Enum):
    TP = "TP"
    PH = "PH"
    PS = "PS"


class FlashDrumOp(IUnitOp):
    def __init__(self, spec: FlashSpec, fc: FlashCalculator,
                 spec_value: float, P: float):
        """
        spec_value is T for TP, H for PH, S for PS.
        P is always specified.
        """
        self._spec = spec
        self._fc = fc
        self._spec_value = spec_value
        self._P = P
        self.feed = Stream()
        self.vapor = Stream()
        self.liquid = Stream()
        self.duty: float = 0.0

    def inlet_ports(self):
        return ["feed"]

    def outlet_ports(self):
        return ["vapor", "liquid"]

    def set_inlet(self, port: str, stream: Stream) -> None:
        if port != "feed":
            raise ValueError(f"FlashDrumOp: unknown inlet port '{port}'")
        self.feed = stream

    def get_outlet(self, port: str) -> Stream:
        if port == "vapor":
            return self.vapor
        if port == "liquid":
            return self.liquid
        raise ValueError(f"FlashDrumOp: unknown outlet port '{port}'")

    def solve(self) -> None:
        if self.feed.total_flow <= 0.0:
            raise ValueError("FlashDrumOp: feed.total_flow must be > 0")

        r_in = self._fc.flash_TP(self.feed.T, self.feed.P, self.feed.z)
        H_in = self._fc.total_enthalpy(r_in)

        if self._spec == FlashSpec.TP:
            r = self._fc.flash_TP(self._spec_value, self._P, self.feed.z)
        elif self._spec == FlashSpec.PH:
            r = self._fc.flash_PH(self._P, self._spec_value, self.feed.z, self.feed.T)
        else:
            r = self._fc.flash_PS(self._P, self._spec_value, self.feed.z, self.feed.T)

        self._populate_outlets(r)
        H_out = self._fc.total_enthalpy(r)
        self.duty = self.feed.total_flow * (H_out - H_in)

    def _populate_outlets(self, r: FlashResult) -> None:
        F = self.feed.total_flow

        self.vapor = Stream()
        self.vapor.T = r.T; self.vapor.P = r.P
        self.vapor.vapor_fraction = 1.0; self.vapor.phase = Phase.VAPOR
        self.vapor.z = self.vapor.y = self.vapor.x = list(r.y)

        self.liquid = Stream()
        self.liquid.T = r.T; self.liquid.P = r.P
        self.liquid.vapor_fraction = 0.0; self.liquid.phase = Phase.LIQUID
        self.liquid.z = self.liquid.x = self.liquid.y = list(r.x)

        self.vapor.total_flow  = F * r.beta
        self.liquid.total_flow = F * (1.0 - r.beta)

        self.vapor.H  = self._fc.phase_enthalpy(r.T, r.P, r.y, False)
        self.liquid.H = self._fc.phase_enthalpy(r.T, r.P, r.x, True)
        self.vapor.S  = self._fc.phase_entropy(r.T, r.P, r.y, False)
        self.liquid.S = self._fc.phase_entropy(r.T, r.P, r.x, True)

        if r.beta >= 1.0 - 1e-10:
            self.vapor.total_flow  = F
            self.liquid.total_flow = 0.0
            self.liquid.z = self.liquid.x = self.liquid.y = list(r.y)
        elif r.beta <= 1e-10:
            self.liquid.total_flow = F
            self.vapor.total_flow  = 0.0
            self.vapor.z = self.vapor.y = self.vapor.x = list(r.x)
